File: Data Structures & Algorithms/trapping-rain-water/submission-7.java
class Solution {
    public int trap(int[] height) {
        // [0,2,0,3,1,0,1,3,2,1]
        
        // [0,2,2,3,3,3,3,3,3,3]
        // [3,3,3,3,3,3,3,3,2,1]
        // [0,0,2,0,2,3,2,0,0,0] = 9
        int n = height.length;
        int [] left = new int[n];
        int [] right = new int [n];
        int result = 0;

        left[0] = height[0];
        right[n -1] = height[n-1];
       

        for(int i = 1; i < height.length; i++){
            left[i] = Math.max(height[i], left[i - 1]);
        }

        for(int i = height.length -2; i >=0; i--){
            right[i] = Math.max(height[i], right[i + 1]);
        }

        for(int i = 0; i < n; i++){
            System.out.println("[" + left[i] + "]" + "[" + right[i] + "]");
            result += Math.min(left[i], right[i]) - height[i];
        }

        return result;
    }
}

File: Data Structures & Algorithms/eating-bananas/submission-7.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        max_value = 0
        result = float('inf')

        for pile in piles:
            max_value = max(pile, max_value)
        
        left = 1
        right = max_value

        while left <= right:
            middle = (left + right) // 2

            sum = 0
            for pile in piles:
                sum += math.ceil(pile / middle)
            

            if sum <= h:
                result = min(middle, result)
                right = middle - 1
            else:
                left = middle + 1
        
        return result

